### PR TITLE
Fix jumping composer when entering emojis or markdown text

### PR DESCRIPTION
**Pull request by Skalakid**
### Fixed Issues

$ https://github.com/Expensify/App/issues/39267
PROPOSAL:

### Tests

1. Go to staging.new.expensify.com/
2 .Go to any chat
3. Enter an emoji and verify composer layout
5. Verify the same when entering other markdown text, pasting text or when expanding composer 
- [X] Verify that no errors appear in the JS console

### Offline tests

Same as tests
### QA Steps

Same as tests
- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [X] I verified that all the inputs inside a form are aligned with each other.
    - [X] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/39538890/4cdaba3c-13b1-47a8-9be8-14cd731c8e19

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/39538890/d1639f78-a309-4132-a295-47e122aa4c4d

</details>

### CODE REVIEW

File: ios/Podfile.lock [hunk 1]
```diff
@@ -1921,8 +1921,8 @@ SPEC CHECKSUMS:
   SocketRocket: f32cd54efbe0f095c4d7594881e52619cfe80b17
   Turf: 13d1a92d969ca0311bbc26e8356cca178ce95da2
   VisionCamera: 9f26224fce1233ffdad9fa4e56863e3de2190dc0
-  Yoga: e64aa65de36c0832d04e8c7bd614396c77a80047
+  Yoga: 13c8ef87792450193e117976337b8527b49e8c03
```
Comment (BartoszGrajdek) [inline ios/Podfile.lock:1924]: They will ask you to remove this upgrade

File: ios/Podfile.lock [hunk 2]
```diff
@@ -1921,8 +1921,8 @@ SPEC CHECKSUMS:
...
 
 PODFILE CHECKSUM: a431c146e1501391834a2f299a74093bac53b530
 
-COCOAPODS: 1.13.0
+COCOAPODS: 1.15.2
```
Comment (BartoszGrajdek) [inline ios/Podfile.lock:1928]: Same here

File: src/components/Composer/index.tsx [hunk 3]
```diff
@@ -237,39 +231,24 @@ function Composer(
     );
 
     /**
-     * Check the current scrollHeight of the textarea (minus any padding) and
-     * divide by line height to get the total number of rows for the textarea.
+     * Check the current text input width and update the state with the new width.
      */
-    const updateNumberOfLines = useCallback(() => {
+    const updateTextInputWidth = useCallback(() => {
```
Comment (BartoszGrajdek) [inline src/components/Composer/index.tsx:236]: The naming seems confusing here, since we also update the height

**Review by BartoszGrajdek — COMMENTED** [review]
Overall seems fine, let me go through everything and see if we can simplify some more 😄

File: src/components/Composer/index.tsx (hunk 3 above)
Comment (Skalakid) [reply src/components/Composer/index.tsx:236]: I removed height updates, so now we are only updating width. I will investigate if we can remove this whole function

**Review by fedirjh — COMMENTED** [review]
Bug: Composer layout shrinks on initial report load on android


https://github.com/Expensify/App/assets/36869046/913e5dd2-f4a4-4181-be93-2a507262b891

Screenshot 2024-04-17 at 6 48 25 AM

---

Bug : Emoji is cropped on the top, Emoji is displayed as large when the comment includes only emojis.

https://github.com/Expensify/App/assets/36869046/0a85cb49-818f-4c82-bff6-19dd125caa69

**Comment by Skalakid** [comment]
@fedirjh thanks for the review. This PR is changing only Composer component for web. Both of the bugs you found are also present on main.

> Bug : Emoji is cropped on the top ...

This one is already reported, https://github.com/Expensify/App/issues/15734

> ... Emoji is displayed as large when the comment includes only emojis.

This is a feature that was implemented in https://github.com/Expensify/App/pull/39597

File: src/components/Composer/index.tsx [hunk 4]
```diff
@@ -333,7 +289,13 @@ function Composer(
                 opacity: 0,
             }}
         >
-            <Text style={[StyleSheet.flatten([style, styles.noSelect]), numberOfLines < maxLines ? styles.overflowHidden : {}, {maxWidth: textInputWidth as DimensionValue}]}>
+            <Text
+                style={[
+                    StyleSheet.flatten([style, styles.noSelect]),
+                    (textInput.current?.clientHeight ?? 0) < 330 ? styles.overflowHidden : {},
```
[the remark this answers is not in the document]
Comment (Skalakid) [reply src/components/Composer/index.tsx]: Changed, I think we can replace it with `maxHeight` instead of setting `overflow: hidden` conditionally

**Comment by Skalakid** [comment]
@fedirjh bump

**Review by fedirjh — COMMENTED** [review]
@Skalakid It seems like the composer is missing a padding. th composer looks smaller compared to production.

| Prod | PR |
|--------|--------|
| Screenshot 2024-04-22 at 7 39 58 AM | Screenshot 2024-04-22 at 7 40 14 AM |
| Screenshot 2024-04-22 at 7 37 17 AM | Screenshot 2024-04-22 at 7 37 03 AM |

**Comment by Skalakid** [comment]
@fedirjh The problem with composer padding was fixed. Also, I removed some height calculations that were based on number of lines, since every line can have different `lineHeight`

**Comment by shawnborton** [comment]
It looks like the emojis are cut off just slightly at the top? 
CleanShot 2024-04-22 at 12 45 46@2x

**Comment by Skalakid** [comment]
> It looks like the emojis are cut off just slightly at the top?

@shawnborton I can't reproduce it, emojis seems to have correct size. However the issue with cutting off emojis at the top is known and reported [here](https://github.com/Expensify/App/issues/40535)

**Review by fedirjh — APPROVED** [review]

**Review by thienlnam — DISMISSED** [review]

**Comment by thienlnam** [comment]
Ah looks like we have some conflicts to resolve now

**Comment by Skalakid** [comment]
I resolved the conflicts

**Review by thienlnam — APPROVED** [review]

**Pull request merged**